homeassistant/components/prometheus/__init__.py

```diff
@@ -221,31 +251,58 @@ def _handle_climate(self, state):
             pass
 
     def _handle_sensor(self, state):
-
         unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-        metric = state.entity_id.split(".")[1]
-
-        if '_' not in str(metric):
-            metric = state.entity_id.replace('.', '_')
+        metric = self._component_config.get(state.entity_id).get(
+            CONF_OVERRIDE_METRIC)
+        if metric in (None, ''):
+            if self._override_metric:
+                metric = self._override_metric
+            else:
+                metric = state.attributes.get(ATTR_DEVICE_CLASS)
+                if metric in (None, ''):
+                    if self._default_metric:
+                        metric = self._default_metric
+                    else:
+                        if unit in (None, ''):
+                            _LOGGER.debug(
+                                "Unsupported sensor: %s",
+                                state.entity_id)
+                            metric = None
+                        else:
+                            metric = "{}_{}".format("sensor_unit",
+                                                    self._unit_string(unit))
+                else:
+                    metric = "{}_{}".format(metric,
```

Comment:
I think we can make this logic block easier to read by extracting the different metric cases into individual handler methods, registering them in our `Registry` decorator util and then looping over the registered handlers calling them and passing the `state`. Each handler should return the metric. We check the return value for `None` and empty string and `break` if not true.

https://github.com/home-assistant/home-assistant/blob/c5239c6176d7a8acf1eeeb29026baea944624ac2/homeassistant/util/decorator.py#L7

```py
METRIC_HANDLERS = Registry()
...

@METRIC_HANDLERS('comp_config')
def handle_comp_config(self, state):
    return self._component_config.get(state.entity_id).get(
        CONF_OVERRIDE_METRIC)

... 
def _handle_sensor(self, state):
    for handler in METRIC_HANDLERS.values():
        metric = handler(state)
        if metric not in (None, ''):
            break
```
We don't really need the decorator registry, if the handlers are instance methods. We could also map the handlers in a dictionary in init method, or save them in a list of we don't need to look up. But I want to mention this tool if there are further improvements possible.

Reply:
Maybe a simple list to store the handlers is best, since we want to use them in order?